src/Grpc.Net.Client/Balancer/Internal/ConnectionManager.cs

```diff
@@ -0,0 +1,467 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Net.Client.Balancer.Internal;
+using Grpc.Net.Client.Configuration;
+using Grpc.Net.Client.Internal;
+using Grpc.Shared;
+using Microsoft.Extensions.Logging;
+
+namespace Grpc.Net.Client.Balancer.Internal
+{
+    internal class ConnectionManager : IDisposable, IChannelControlHelper
+    {
+        private readonly SemaphoreSlim _nextPickerLock;
+        private readonly object _lock;
+        private readonly Resolver _resolver;
+        private readonly ISubchannelTransportFactory _subchannelTransportFactory;
+        private readonly List<Subchannel> _subchannels;
+        private readonly List<StateWatcher> _stateWatchers;
+        private readonly CancellationTokenSource _cts;
+
+        // Internal for testing
+        internal LoadBalancer? _balancer;
+        internal SubchannelPicker? _picker;
+        private Task? _resolverRefreshTask;
+        private Task? _resolveTask;
+        private TaskCompletionSource<SubchannelPicker> _nextPickerTcs;
+
+        internal ConnectionManager(
+            Resolver resolver,
+            bool disableResolverServiceConfig,
+            ILoggerFactory loggerFactory,
+            ISubchannelTransportFactory subchannelTransportFactory,
+            LoadBalancerFactory[] loadBalancerFactories)
+        {
+            _lock = new object();
+            _nextPickerLock = new SemaphoreSlim(1);
+            _nextPickerTcs = new TaskCompletionSource<SubchannelPicker>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _cts = new CancellationTokenSource();
+
+            Logger = loggerFactory.CreateLogger(GetType());
+
+            _subchannels = new List<Subchannel>();
+            _stateWatchers = new List<StateWatcher>();
+            _resolver = resolver;
+            DisableResolverServiceConfig = disableResolverServiceConfig;
+            _subchannelTransportFactory = subchannelTransportFactory;
+            LoadBalancerFactories = loadBalancerFactories;
+        }
+
+        public ConnectivityState State { get; private set; }
+        public ILogger Logger { get; }
+        public bool DisableResolverServiceConfig { get; }
+        public LoadBalancerFactory[] LoadBalancerFactories { get; }
+
+        public IList<Subchannel> GetSubchannels()
+        {
+            lock (_subchannels)
+            {
+                return _subchannels.ToArray();
```

Comment:
Is this redundant or are we making a copy to avoid mutating the _subchannels stored here? Also, where was this used?

Reply:
This collection could change at any time. This method is just for unit tests.